Лаб. 5 6411 Шехматов Виктор

Nakamo312 opened the pull request:
**ФИО:** Шехматов Виктор  
**Номер группы:** 6411 
**Номер лабораторной:** 5  
**Номер варианта:** 25 
**Краткое описание предметной области:**Клиника.
**Краткое описание добавленных фич:** Добавлено клиентское приложение на Blazor WebAssembly. Настроен HttpClient для работы с API. Добавлена конфигурация JSON сериализации с поддержкой enum. Настроен CORS между клиентом и API. Реализованы базовые UI компоненты для работы с пациентами, врачами, записями и специализациями. Добавлена интеграция с Aspire для автоматической конфигурации endpoint.

Nakamo312 committed:
initial commit

Nakamo312 committed:
fix test.yml

Nakamo312 committed:
add comments for IModel.cs

Nakamo312 committed:
change field IsRepeated in Appointment data model

Nakamo312 committed:
added Repository class. Added DataBase context class

Nakamo312 committed:
create Application Layer: -Added MappingProfile;\n-Added DTOs;

Nakamo312 committed:
created Base Controller and Services

Nakamo312 committed:
modified DTOs

Nakamo312 committed:
Add API Controllers

Nakamo312 committed:
Add InMmempry Repositories

Nakamo312 committed:
Added API CRUD endpoints, added DTO representation schemes, addrd logging through middleware

Nakamo312 committed:
update README.md

Nakamo312 committed:
changes status code and response header in Create methods

Nakamo312 committed:
Added analytic services and controllers

Nakamo312 committed:
rm file

Nakamo312 committed:
fix test.yml

Nakamo312 committed:
fix test.yml

Nakamo312 committed:
remove exceptions messages from endpoints

Nakamo312 committed:
change swager gen documentation for DTO schemes. Change content-type for all endpoints

Nakamo312 committed:
refactor: decouple API layer from domain models

- Remove Domain.Model dependency from base controller
- Use reflection for ID access in DTOs instead of Model inheritance
- Keep API layer dependent only on DTOs and application contracts

Nakamo312 committed:
Remarks have been fixed

Nakamo312 committed:
Remarks have been fixed

Nakamo312 committed:
Remarks have been fixed

Nakamo312 committed:
Remarks have been fixed

Nakamo312 committed:
fix regex in PatientUpdateDto

Nakamo312 committed:
Improve error management: hide detailed validation errors

Nakamo312 committed:
revert last changes

Nakamo312 committed:
change response stats code in DELETE methods

Nakamo312 committed:
Aspire mood orchestrator; creating a conference for the database

Nakamo312 committed:
update README.MD

Nakamo312 committed:
comments corrected

Nakamo312 committed:
rename folder

Nakamo312 committed:
rename folder

Nakamo312 committed:
rename folder

Nakamo312 committed:
rename folder

Nakamo312 committed:
rename folder

Nakamo312 committed:
rename folder

Nakamo312 committed:
registred DataSeed in DI container

Nakamo312 committed:
update phone number regex

Nakamo312 committed:
added phone number normalizatiom; change contracts for PUT methods

Nakamo312 committed:
changes namespaces

Nakamo312 committed:
changes namespaces

Nakamo312 committed:
modified README.md

Nakamo312 committed:
change Domain Models type field from uint to Guid

Nakamo312 committed:
refactor Program.cs

Nakamo312 committed:
Rename tablenames and columns

Nakamo312 committed:
s

Nakamo312 committed:
Add project ClinicDataGenerator, Clinic.RabbitMqProducer,Clinc.RabbitMqConsumer

Nakamo312 committed:
Rename appsettingsDevelopment.json to appsettings.Development.json

Nakamo312 committed:
Fix code review issues

Nakamo312 committed:
Merge branch 'lab4' of https://github.com/Nakamo312/enterprise-development into lab4

Nakamo312 committed:
update reference

Nakamo312 committed:
Added Blazor WebAssembly client application

danlla commented on Clinic/Clinic.Client/Program.cs:
```diff
@@ -0,0 +1,17 @@
+builder.RootComponents.Add<HeadOutlet>("head::after");
+builder.Services.AddScoped(sp => new HttpClient
+{
+    BaseAddress = new Uri(builder.Configuration["ServiceDiscovery:Endpoints:clinic-api"])
```
> Possible null reference argument for parameter 'uriString' in 'Uri.Uri(string uriString)'

При запуске падает здесь

danlla commented on Clinic/Clinic.AppHost/AppHost.cs:
```diff
@@ -0,0 +1,33 @@
+
+builder.AddProject<Projects.Clinic_Client>("clinic-client")
+    .WithReference(api)
+    .WithExternalHttpEndpoints();
```
Лучше здесь в клиент прокинуть путь до сервера через переменные окружения (`Api__BaseUrl`), можно использовать `api.GetEndpoint`

danlla replied:
Здесь не хватает подождать `api`

danlla commented on Clinic/Clinic.Client/Program.cs:
```diff
@@ -0,0 +1,17 @@
+builder.RootComponents.Add<HeadOutlet>("head::after");
+builder.Services.AddScoped(sp => new HttpClient
+{
+    BaseAddress = new Uri(builder.Configuration["ServiceDiscovery:Endpoints:clinic-api"])
```
И тогда использовать это значение здесь

Nakamo312 replied:
Я понятия не имею как это сделать без генерации json в www-root. Я предполагаю что в WASM нет прямого доступа к переменным среды сервера

Nakamo312 replied:
<img width="751" height="770" alt="изображение" src="https://github.com/user-attachments/assets/3062b57e-6b7c-46d5-98d8-c37ab073d195" />
в Aspire видно что переменная есть

```C#
var apiBaseUrl = builder.Configuration["Api__BaseUrl"] 

builder.Services.AddScoped(sp => new HttpClient
{
    BaseAddress = new Uri(apiBaseUrl)
});
``` 

но в браузере вижу что запросы идут не к API а к baseURI самого клиента

danlla replied:
`appsettings.json` должен быть в `wwwroot`

Nakamo312 replied:
Да, это помогло. Есть теперь другая проблема - я сколкьо бы не сидел не втыкал как работают очереди
ошибка ```MessageText: insert or update on table "appointments" violates foreign key constraint "FK_appointments_patients_patient_id"
    Detail: Detail redacted as it may contain sensitive data. Specify 'Include Error Detail' in the connection string to include this information.
    SchemaName: public
    ``` не прилетает просто так. Я понял что она появляется когда удаляешь сущности - которые EntityTrackerId должен по хорошщему убирать из зарегестрированных генератором.
    там адские танцы с бубном пошли чтоб как то это обрабатывать и я вприницпе пожалел что решил генерить контракты в цикле и хранить в оперативной памяти id созданых сущностей для остлеживания. Боюсь времени не переделку не хватает

Nakamo312 replied:
я думаю самое оптимально что я успею это просто обрабатывать это исключение и удалять из трекера такие айди чтоб генератор больше не использовал их

danlla replied:
Сделайте по минимуму просто, предположим сущности не будут удаляться, поправьте остальные замечания

danlla commented on Clinic/Clinic.Client/appsettings.json:
```diff
@@ -0,0 +1,5 @@
+{
+  "Api": {
+    "BaseUrl": "https://localhost:7000"
+  }
```
Нигде же не используется

danlla commented on Clinic/Clinic.Client/Program.cs:
```diff
@@ -0,0 +1,17 @@
+using Microsoft.AspNetCore.Components.Web;
+using Microsoft.AspNetCore.Components.WebAssembly.Hosting;
+using System.Text.Json; 
+using System.Text.Json.Serialization; 
```
cleanup

danlla commented on Clinic/Clinic.AppHost/AppHost.cs:
```diff
@@ -0,0 +1,33 @@
+﻿using Google.Protobuf.WellKnownTypes;
```
cleanup

danlla commented on Clinic/Clinic.Client/wwwroot/sample-data/weather.json:
Этот файл можно удалить

danlla commented on Clinic/Clinic.Client/Properties/launchSettings.json:
```diff
@@ -0,0 +1,41 @@
+      "sslPort": 44309
+    }
+  },
+  "profiles": {
+    "http": {
+      "commandName": "Project",
+      "dotnetRunMessages": true,
+      "launchBrowser": true,
+      "inspectUri": "{wsProtocol}://{url.hostname}:{url.port}/_framework/debug/ws-proxy?browser={browserInspectUri}",
+      "applicationUrl": "http://localhost:5115",
+      "environmentVariables": {
+        "ASPNETCORE_ENVIRONMENT": "Development"
+      }
+    },
+    "https": {
+      "commandName": "Project",
+      "dotnetRunMessages": true,
+      "launchBrowser": true,
+      "inspectUri": "{wsProtocol}://{url.hostname}:{url.port}/_framework/debug/ws-proxy?browser={browserInspectUri}",
+      "applicationUrl": "https://localhost:7145;http://localhost:5115",
+      "environmentVariables": {
+        "ASPNETCORE_ENVIRONMENT": "Development"
+      }
+    },
+    "IIS Express": {
+      "commandName": "IISExpress",
+      "launchBrowser": true,
+      "inspectUri": "{wsProtocol}://{url.hostname}:{url.port}/_framework/debug/ws-proxy?browser={browserInspectUri}",
+      "environmentVariables": {
+        "ASPNETCORE_ENVIRONMENT": "Development"
```
Вместе с дашбордом aspire открывается браузер с клиентом, что не очень удобно
Здесь можно проставить `"launchBrowser": false`, чтобы браузер с клиентом не открывался 

Возможно это можно настроить как-то еще

danlla commented on Clinic/Clinic.Client/appsettings.json:
Чтобы `appesttings.json` подгружался, он должен быть в папке `wwwroot`

danlla reviewed (changes requested):
В ридми можно написать про все лабораторные, про взаимодействие приложений между собой

В данном ревью замечания только к 5 лабораторной, с 4 еще осталось то, что нужно поправить

Использование готовых компонентов сильно упростило бы жизнь

Со временем начинает валиться много ошибок вида:

```
Microsoft.EntityFrameworkCore.DbUpdateException: An error occurred while saving the entity changes. See the inner exception for details.
 ---> Npgsql.PostgresException (0x80004005): 23503: insert or update on table "appointments" violates foreign key constraint "FK_appointments_patients_patient_id"

DETAIL: Detail redacted as it may contain sensitive data. Specify 'Include Error Detail' in the connection string to include this information.
   at Npgsql.Internal.NpgsqlConnector.ReadMessageLong(Boolean async, DataRowLoadingMode dataRowLoadingMode, Boolean readingNotifications, Boolean isReadingPrependedMessage)
   at System.Runtime.CompilerServices.PoolingAsyncValueTaskMethodBuilder`1.StateMachineBox`1.System.Threading.Tasks.Sources.IValueTaskSource<TResult>.GetResult(Int16 token)
   at Npgsql.NpgsqlDataReader.NextResult(Boolean async, Boolean isConsuming, CancellationToken cancellationToken)
   at Npgsql.NpgsqlDataReader.NextResult(Boolean async, Boolean isConsuming, CancellationToken cancellationToken)
   at Npgsql.NpgsqlCommand.ExecuteReader(Boolean async, CommandBehavior behavior, CancellationToken cancellationToken)
   at Npgsql.NpgsqlCommand.ExecuteReader(Boolean async, CommandBehavior behavior, CancellationToken cancellationToken)
   at Npgsql.NpgsqlCommand.ExecuteDbDataReaderAsync(CommandBehavior behavior, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReaderAsync(RelationalCommandParameterObject parameterObject, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReaderAsync(RelationalCommandParameterObject parameterObject, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Update.ReaderModificationCommandBatch.ExecuteAsync(IRelationalConnection connection, CancellationToken cancellationToken)
  Exception data:
    Severity: ERROR
    SqlState: 23503
    MessageText: insert or update on table "appointments" violates foreign key constraint "FK_appointments_patients_patient_id"
    Detail: Detail redacted as it may contain sensitive data. Specify 'Include Error Detail' in the connection string to include this information.
    SchemaName: public
    TableName: appointments
    ConstraintName: FK_appointments_patients_patient_id
    File: ri_triggers.c
    Line: 2599
    Routine: ri_ReportViolation
   --- End of inner exception stack trace ---
   at Microsoft.EntityFrameworkCore.Update.ReaderModificationCommandBatch.ExecuteAsync(IRelationalConnection connection, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.ExecuteAsync(IEnumerable`1 commandBatches, IRelationalConnection connection, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.ExecuteAsync(IEnumerable`1 commandBatches, IRelationalConnection connection, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.ExecuteAsync(IEnumerable`1 commandBatches, IRelationalConnection connection, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Storage.RelationalDatabase.SaveChangesAsync(IList`1 entries, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChangesAsync(IList`1 entriesToSave, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChangesAsync(StateManager stateManager, Boolean acceptAllChangesOnSuccess, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Storage.ExecutionStrategy.<>c__DisplayClass30_0`2.<<ExecuteAsync>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at Microsoft.EntityFrameworkCore.Storage.ExecutionStrategy.ExecuteImplementationAsync[TState,TResult](Func`4 operation, Func`4 verifySucceeded, TState state, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Storage.ExecutionStrategy.ExecuteImplementationAsync[TState,TResult](Func`4 operation, Func`4 verifySucceeded, TState state, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Storage.ExecutionStrategy.ExecuteAsync[TState,TResult](TState state, Func`4 operation, Func`4 verifySucceeded, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.DbContext.SaveChangesAsync(Boolean acceptAllChangesOnSuccess, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.DbContext.SaveChangesAsync(Boolean acceptAllChangesOnSuccess, CancellationToken cancellationToken)
   at Clinic.Infrastructure.Repositories.Repository`1.CreateAsync(T entity) in Clinic\Clinic.Infrastructure\Repositories\Repository.cs:line 23
   at Clinic.RabbitMq.Consumer.Consumers.RabbitMqConsumer.ProcessAppointment(IServiceScope scope, Object payload) in Clinic\Clinic.RabbitMq.Consumer\Consumers\RabbitMqConsumer.cs:line 266
   at Clinic.RabbitMq.Consumer.Consumers.RabbitMqConsumer.ProcessEntity(String queue, Object payload) in Clinic\Clinic.RabbitMq.Consumer\Consumers\RabbitMqConsumer.cs:line 184
   at 
Clinic.RabbitMq.Consumer.Consumers.RabbitMqConsumer.ProcessMessageAsync(String queueName, String entityType, Byte[] body) in Clinic\Clinic.RabbitMq.Consumer\Consumers\RabbitMqConsumer.cs:line 157
   at Clinic.RabbitMq.Consumer.Consumers.RabbitMqConsumer.<>c__DisplayClass11_0.<<StartConsumersAsync>b__0>d.MoveNext() in Clinic.RabbitMq.Consumer\Consumers\RabbitMqConsumer.cs:line 116
```

Nakamo312 committed:
refactor(client): integrate MudBlazor components and fix RabbitMQ entity tracking

- Complete UI overhaul using MudBlazor for modern component-based design
- Fixed critical bug in RabbitMQ communication where deleted entities caused failures
- Resolved all issues from previous code review and improved component architecture

Nakamo312 commented:
Вобще главную страницу можно было на вебсокетах сделать чтоб реальные данные трекать, но это просто заглушка, причем завайбкожена, просьба не судить строго
